Pyrado/pyrado/utils/bijective_transformation.py

```diff
@@ -0,0 +1,135 @@
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+# 3. Neither the name of Fabio Muratore, Honda Research Institute Europe GmbH,
+#    or Technical University of Darmstadt, nor the names of its contributors may
+#    be used to endorse or promote products derived from this software without
+#    specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL FABIO MURATORE, HONDA RESEARCH INSTITUTE EUROPE GMBH,
+# OR TECHNICAL UNIVERSITY OF DARMSTADT BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+# IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import math
+from abc import ABC, abstractmethod
+from typing import Union
+
+import numpy as np
+import torch as to
+
+
+class BijectiveTransformation(ABC):
+    """
+    Base class for bijective transformations to be used for e.g. transforming the domain parameter space of an env.
+
+    These transformations are useful for avoiding infeasible values such as negative masses. The reasoning behind this
+    is that some learning methods work on the set of real numbers, thus we make them learn in the transformed space,
+    here the log-space, without telling them.
+    """
+
+    @abstractmethod
+    def forward(self, value: Union[int, float, np.ndarray, to.Tensor]) -> Union[int, float, np.ndarray, to.Tensor]:
+        """
+        Map a value from the space to the transformed space.
+
+        :param value: value in the original space
+        :return: value in the transformed space
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def inverse(self, value: Union[int, float, np.ndarray, to.Tensor]) -> Union[int, float, np.ndarray, to.Tensor]:
+        """
+        Map a value from the transformed space to the actual space.
+
+        :param value: value in the transformed space
+        :return: value in the original space
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def ensures_non_negativity(self) -> bool:
+        """
+        Whether this transformations ensures non-negativity, i.e., whether :py:meth:`.inverse` only returns non-negative
+        values.
+        """
+        raise NotImplementedError()
```

Comment:
Is this really necessary? I mean ppl should know this, right?

Reply:
This was used by the sprl implementation to decide whether a non-negativity constraint has to be applied or not. I'd leave it in 'cause it doesn't hurt.